operations/iam/openstack-federation-via-oidc.md

```diff
@@ -0,0 +1,98 @@
+in SCS.
+
+The following sections describe the setup.
+
+## 1. Keystone
+
+[Keytone supports federated identities](https://docs.openstack.org/keystone/latest/admin/federation/federated_identity.html).
+To allow SCS to consume identities managed in external IAM solutions,
+federation protocols like OpenID Connect or SAML can be used.
+Keystone currently makes use of third party apache modules like
+[mod\_auth\_openidc](https://github.com/OpenIDC/mod_auth_openidc),
+[mod\_oauth2](https://github.com/OpenIDC/mod_oauth2) and
+[mod\_auth\_mellon](https://github.com/UNINETT/mod_auth_mellon) to delegate
+authentication to an SSO IdP (i.e. SAML IdP or OpenID Connect provider).
+
+In OpenStack the apache modules are configured using the
+[wsgi-keystone.conf template](https://github.com/openstack/kolla-ansible/tree/master/ansible/roles/keystone/templates).
+
+In SCS we make use of the OAuth 2.0 Authorization Code Grant flow between Keystone and Keycloak
+and use PKCE (RFC 7636) with the S256 challenge method.
+
+In addition to the usual SSL CA of the environment, Keycloak uses separate certificates to sign the OIDC tokens.
+
+Due to the way the Keystone container image runs apache (in the foreground) and keystone itself (as WSGI module),
+reconfiguring the apache URL locations on the fly is not possible currently without a downtime of several seconds.
+That is the reason why SCS currently makes use of a single central proxy realm in Keycloak, to which Keystone
+connects.
+
+### 1.1 Keycloak IdP realm discovery
+
+Keycloak offers standard OIDC service discovery via `.well-known` documents to advertise its endpoints.
+
+In SCS we want to represent each customer by a separate dedicated Keycloak realm, which can enventually be used
+for customer self service and to federate out to customer owned IAM external solutions.
+
+In the SCS testbed we currently experiment with the implications of using a single central proxy realm in Keycloak
+and chaining federation from there to customer specific realms, also hosted in the same Keycloak instance.
+To make this usable, SCS makes use of the
+[Keycloak Home IdP Discovery](https://github.com/sventorben/keycloak-home-idp-discovery)
+extension.
+
+### 1.2 Keystone mapping of token claims
+
+Upon login of a user Keystone evaluates the credentials obtained from the ID token that the IdP issued.
+These include group memberships and roles, which can be used to assign the user to a certain project.
+Keystone maps these external identities to internal (shadow) users.
+It can either attempt to map the obtained information to a `local` type user, which needs to be
+provisioned before authentication by external tooling, or it can be instructed via the mapping to
+generate an `ephemeral` type user. `ephemeral` users are cleaned up automatically after some time
+of inactivity and with that, they lose access to projects, iff the access is granted indirectly
```

Comment:
```suggestion
of inactivity and with that, they lose access to projects, if the access is granted indirectly
```

Reply:
No, I wanted `iff`, in the mathematical sense: If and only if.